### ISSUE TITLE
Does folder containing tests have to be called "tests"?

### ISSUE THREAD

**Issue by yonigibbs** [klazuka/intellij-elm#563]
If the folder containing tests isn't named `tests`, the code in IntelliJ goes red. Is there a setting I need to set somewhere?

This is shown in the screenshot below, where the folder is called `test` rather than `tests`:

![elm test folder](https://user-images.githubusercontent.com/39593145/70629431-e635b780-1c21-11ea-8d34-33fa6ad190f8.png)

If I rename the folder from `test` to `tests`, the red code goes black as normal.

The screenshot above is just a cut down example, where renaming the folder to "tests" would obviously be an easy answer.  However the reason I don't want a folder called "tests" is because in a real solution the structure is a bit more complicated (it's a Spring Boot app):

```
- src
  - main
    - kotlin
    - elm
  - test
    - kotlin
    - elm
      - <tests go here>
```

Note that I can run elm-test on a command-line and the tests run fine, regardless of the name of the folder.

**Comment by yonigibbs** [comment]
I had a dig around the source code and I can see that this value ("tests") is hard-coded in a few places, e.g. `ElmProject`, `ElmModuleBuilder`, etc.  So looks like as the code currently stands the folder does indeed need to be called "tests".

I can also see that there is this comment in `ElmWorkspaceService`:

> Testing on Elm 0.19 requires that there be a "tests" directory next to the elm.json file. So we must register that as a source root if it exists.

Is that definitely true?  With the elm-test CLI you can specify the path to the test files, and that seems to work even if they're not in a folder called "tests".

If you agree it is possible for elm-test to work in a setup without a "tests" directory (which is how we've got it set up, and haven't noticed any problems so far), would it be possible to make the path to the folder containing the tests configurable?  Maybe a new value the user can enter in the Settings screen where they enter the location of elm-test?

**Comment by klazuka** [comment]
@yonigibbs which version of elm-test are you using? I'm using 0.19.1-revision2, and I do not see an option to configure the location of the tests. When I rename my `tests` directory to, say, `foo`, and try to run `elm-test foo/*.elm`, I get the following error:
```
Error: /Users/keith/dev/elm-json-tree-view/tests does not exist. Please create a tests/ directory in your project root!
```

This makes it seem that elm-test only supports a directory named "tests".

**Comment by yonigibbs** [comment]
I'm using the same version as you. I had to add an entry in `elm.json` under the `source-directories` key, e.g. if my tests are in `my-tests` then I have to do this:

```
{
    "type": "application",
    "source-directories": [
        "src", "my-tests"
    ],
    "elm-version": "0.19.1",
    ...
```

**Comment by klazuka** [comment]
Hm. That's pretty hacky, and I'm not crazy about adding more configuration to the plugin which depends on such an unusual setup. It seems to me that `elm-test`'s ability to specify the path to the source files is intended solely to pick a subset of files from within `tests` to run. Going outside of the `tests` directory seems unintentional.

I was going to suggest that you use a symlink, but `node-test-runner` rejects that, too 🙁 

If elm-test ever formally supports this kind of setup, then I would definitely make the changes to support it. But for now, I'm going to close it. Sorry.

**Comment by yonigibbs** [comment]
No worries at all, @klazuka, that's totally fine with me.  Thanks for looking into this anyway.

**Comment by yonigibbs** [comment]
Actually, just one final tiny point on this, @klazuka: I think the discussion here suggests that the `tests` folder is no longer meant to be a hard limitation, at least for applications rather than packages: https://github.com/rtfeldman/node-test-runner/pull/306.  And looks like other people have looked for this before too: https://stackoverflow.com/questions/52577282/change-tests-directory-in-elm-test.  As you said, the symlink idea was worth a shot, but is rejected by node-test-runner.  I think in an environment where you've got one project which contains front end and back end stuff, enforcing the front-end unit tests to be in a `tests` folder at the top level might not be the best option.

For us it's not a big deal - we can write the tests outside of IntelliJ and run them on the command line, but just wanted to make sure you were aware of that PR and the related discussion.

**Comment by klazuka** [comment]
Oh, I see. I was testing it out with one of my Elm *package* projects, but elm-test only allows non-`tests` directory for *applications*. 

Ideally the test root would be in the `elm.json` file, but I guess we can't hold our breath for that kind of change. Figuring out how to model this may be a bit tricky. Right now our persistent settings are at the "ElmWorkspace" level. A workspace corresponds to one IntelliJ project, and the workspace can contain multiple, attached Elm projects (`elm.json` files).

Maybe the better place to put this would be in the run configuration state associated with the `elm-test` run config. @frawa just rewrote the elm-test integration in Kotlin, and I have an active PR up to make it a little more idiomatic. But once that PR lands, maybe you could look into providing the user the ability to override the "tests" directory in `ElmTestRunConfiguration`. The other thing that would need to be done is to consolidate the hard-coded logic that assumes every Elm 0.19 test is in a directory named `tests` adjacent to the `elm.json` file.

**Comment by yonigibbs** [comment]
Perfect. Ping me when you feel this is ready and I'll take a look then.

**Comment by klazuka** [comment]
@yonigibbs I just merged the PR

**Comment by yonigibbs** [comment]
Just doing some initial investigations, and got some questions... at first I thought that in the plugin's settings screen we'd just let the user choose the folder containing their tests. But as you say that won't work for the case where one IntelliJ project can contain multiple Elm projects. So the next suggestion was to put this in the run configuration: the problem with that is that the run configuration doesn't seem to be created till the user tries to run the tests (e.g. right-clicks on the test file and chooses "Run 'Tests in <project-name>'". And this menu only appears if the system thinks this file is a valid Elm source file. Currently if it's in a folder called "tests" this is fine, otherwise it won't think it's a valid Elm source file unless the folder it's in appears in the `source-directories` value specified in `elm.json`. Previously in this thread we said that was a bit hacky, so do you want to avoid users having to edit `elm.json`? Note that for `elm-test` to run on the cmd line with a folder other than `tests`, you also need to update `elm.json` and add the custom folder into `source-directories`, so it might be that this step is unavoidable (though maybe we can automate it to save the user having to do it manually).

So there seems to be a slight chicken/egg situation. Do you have any thoughts on what the ideal user-experience here would be? One option is that we let them right-click on _any_ `.elm` file and get the "Run Tests..." menu item. When they click that, we check what folder that file is in, check if it's specified in `elm.json` and, if not, add it in. Then we run `elm-test` supplying the path to the chosen file. But this would change the current behaviour where the run configuration runs _all_ tests rather than just the tests in the single file they clicked on.

When you get chance, let me know your thoughts on what the desired functionality is. (I'm on the Elm slack if it's easier to chat there.)

Another couple of (unrelated) things I noticed... when you right-click in a source file and choose "Run Tests..." it creates the run configuration and executes it. If you then repeat the same action, a new run configuration is created and executed every time you do this. This is different from on a Java/Kotlin source file where it will re-use the same run configuration.  And secondly, that run configuration will execute all the tests in your Elm project, not just the ones in the file you've selected.  Again, this is different from with Java/Kotlin source files.  Do you think either of these should be logged as issues?  Or is this behaviour OK?

**Comment by klazuka** [comment]
### Source-direcctory hack

Let's make it the user's responsibility to add it to `source-directories`. Presumably they will have got things working from the command line already. And I would hope that as `elm-test` becomes more deeply integrated with Elm itself, they will make changes to the `elm.json` format to include a setting for test source dir.

However, there may be another way to handle this which has additional advantages. Right now the plugin shows a yellow banner at the top of the editor when it cannot find an associated `ElmProject` (`elm.json`) for the current `.elm` source file. And it gives the user the option to attach another `elm.json` file to the workspace to solve the problem. But in this case, the solution is not to attach but to add the containing directory as a `source-directory` to an *existing* `ElmProject` in the current workspace.

By taking this approach, we can solve this case as well as other cases (i.e. the user creates a `vendor` directory and moves some third-party Elm code into it but forgets to update `source-directories`). The downside might be that by suggesting it to the user, we might lead them down the wrong path in some cases. Wording would be important.

### Duplicate test configs

Re-running the tests via the context menu (or `ctrl-shift-r`) should not re-create the run config. Maybe that's a regression when the code was ported from Java to Kotlin. I will create an issue for it.

### Restricting tests to specific file(s)

Restricting the scope of tests based on context would be nice. I suspect that this would be a bit of work, so I would leave it out-of-scope for now. It looks like 0.18 elm-test only accepts a single file as a parameter, whereas 0.19 accepts multiple.
